i3: add module

uvNikita opened the pull request:
~~Note: this is **work in progress** PR, created merely to identify that the new module is being developed, and to prevent duplicating work.~~
~~While the code is currently in active development, comments, ideas, and feedback on the module API are highly appreciated. Be aware that commits might be amended and rebased in the process.~~

Progress:
- [x] ~~variables~~ (we can just use nix for that)
- [x] keybindings
- [x] autostart (exec and exec_always)
- [x] modes
- [x] fonts
- [x] bar
- [x] assigns
- [x] floating windows
- [x] colors
- [x] borders
- [x] focus settings
- [x] sensible defaults
- [x] reload on config change

uvNikita commented:
I think it's pretty much ready for merge.

One of my goals was to make it possible to enable module via `xsession.windowManagers.i3.enable = true` to get working i3 with the default parameters. I also wanted to give people possibility to set `i3.config = null` and to configure module with existing text configuration using `i3.configExtra` parameter.

My personal configuration that uses almost all options: https://pastebin.com/QTUnEuBX.

Default values were taken from here: https://github.com/i3/i3/blob/next/etc/config.
i3 config documentation: https://i3wm.org/docs/userguide.html

@rycee if I understand correctly you are using xmonad, but maybe you look at this PR from a general perspective? I would appreciate any feedback. For now, I put the code in a separate `xsession.wm.i3` submodule since I didn't want to pollute xsession module with extra stuff. Maybe we could switch to a new way of supporting window managers before merging this pull request (and I will switch form `xsession.wm` to `xsession.windowManager`)? It's almost a month has passed since we deprecated `xsession.windowManager` parameter which should have given users enough time to switch. What do you think?

rycee commented on modules/services/window-managers/i3.nix:
```diff
@@ -0,0 +1,532 @@
+            titlebar = mkOption {
+              type = types.bool;
+              default = cfg.package != pkgs.i3-gaps;
+              defaultText = "cfg.package != pkgs.i3-gaps (titlebar should be disabled for i3-gaps)";
```
The "cfg" here isn't very understandable for a user reading the documentation. Unfortunately will need something more extensive like "xsession.wm.i3.package != pkgs.i3-gaps" (maybe even change "pkgs" to "nixpkgs"?).

uvNikita replied:
Good point!

rycee commented:
Wow! This is quite the module! 😄

I just had a very quick look right now. I'll try to look it over a bit more carefully tomorrow, although as you pointed out I'm an xmonad user so I'll avoid looking at the particulars of i3.

I'll then also add a news entry saying that the `xsession.windowManager` option as a string will be removed completely on October 31 (~1 month after the [deprecation merge](https://github.com/rycee/home-manager/pull/80#issuecomment-333301198)) . Is it OK for you to delay the merge of this module until then?

uvNikita commented:
Thank you!

Yes, I think we can wait until October 31, I think more people will likely to use home-manager if we will not break anything unexpectedly :)

Hopefully, some i3 users will also show-up to test this module in the meantime :)

pmiddend commented:
I tested the module and was delighted to see that I was able to migrate 99% of my config to home-manager. Just one thing: I didn't find the "focus_follows_mouse" option in the i3 module. Is there a way to set this option (its on by default - I like it off).

uvNikita commented:
Thank you for testing it out @pmiddend ! Let me know if you think some option names or submodules structure is confusing, we can still discuss it and change while it's not merged yet.

Good point about the default font; fixed that. I also added `i3.focus` submodule where you can set `followMouse`, `forceWrapping` and `newWindow` parameters.

pmiddend commented:
Works like a charm, thanks!

rycee commented on modules/services/window-managers/i3.nix:
```diff
@@ -0,0 +1,568 @@
+              defaultText = "xsession.wm.i3.package != nixpkgs.i3-gaps (titlebar should be disabled for i3-gaps)";
+              description = "Whether to show window titlebars.";
+            };
+            border = mkOption {
```
Generally I think it's nice to have an empty line between options. You have such separators above but not below.

rycee commented:
I think the module is just fine. The single concern I may raise is that the number of options is so large that it might make the module cumbersome to maintain and the documentation cluttered. But if you are willing to maintain it then I think it is OK to merge :-)

uvNikita commented:
I see your point about a number of options and cluttered documentation. The reason I wrote the module in this way is that for my use-case it is very convenient to have this level of flexibility. At the same time, I was also trying to make this module useful for simpler cases. Since I use i3 on daily bases, I don't mind to maintain this module.

One way I see to simplify the module and significantly reduce the documentation size (by 30%) is to change `i3.config.colors` module to this:
```
colors = mkOption {
  type = types.submodule {
    options = {
      background = mkOption {
        type = types.string;
        default = "#ffffff";
        description = ''
          Background color of the window.
          Only applications which do not cover the whole area expose the color.
        '';
      };

      window = mkOption {
        type = types.attrsOf colorSetModule;
        default = {
          focused = { ... };
          focusedInactive = { ... };
          unfocused = { ... };
          urgent = { ... };
          placeholder = { ... };
        };
        description = "Color settings for different states of windows";
      };
    };
  };
  default = {};
  description = "Colors settings.";
};
```

I see two downsides to this approach:
  1.  We will lose a compile-time check of the attribute names (like `focused`, `urgent` etc).
  2. And possibility for users to specify only the groups that they want to change (like in my config above) and get other values set to defaults. I'm not sure if there is any way to preserve this behavior though. E.g. if we could somehow merge default values with the once specified by users.

I'm personally fine with both of the downsides if you think this would improve the module @rycee.

rycee commented:
About the colors, maybe could do something similar to the vim module? I.e., set `visible = false` on the options in `colorSetModule` and describe the color options in, e.g., `colors.window.focused` and in the other options say something like "See <varname>colors.window.focused</varname> for configuration options".

uvNikita commented:
I made `colorSetModule` fields invisible as you suggested and added more documentation to the color submodule describing how to use it. Thank you for the input!

uvNikita commented:
I squashed commits to prepare for the rebase. Note that the module still using `xsession.wm` path. Once xmonad will be moved to its own submodule and `xsession.windowManager` will be removed, I will modify this PR. I could create PR for the xmonad migration, but since I don't use it, I can't test if it will work as should after changes.

rycee commented:
No worries. I'll switch over the windowManager option shortly and will take care of xmonad at the same time. I'll ping in this PR once it's done :-)

uvNikita commented:
Thank you! :-)

rycee commented:
@uvNikita Ok, now it should be sorted.

uvNikita commented:
Great! I now modified the commit to use `xsession.windowManager` path.

rycee commented:
Looks good. Feel free to merge whenever you are happy with it.

uvNikita commented:
Rebased into master https://github.com/rycee/home-manager/commit/467b774d13ec47e623524d20acd7b19e30927480. Thanks to all who participated in the PR!

rycee commented:
Wonderful work! Thanks for the patience in getting this in 👍

GaetanLepage commented:
Hello @uvNikita 
What is the reason for the initialization of `default_border` and `default_floating_border` ?
Why shouldn't it be `normal` when using i3-gaps ?

uvNikita commented:
@GaetanLepage It's been awhile, I don't really remember why titlebar had to be disabled for i3-gaps...
I vaguely remember that, at the time of writing the module, title bars didn't play well with gaps.
If it has been fixed, it should probably be set to the default value of i3 itself.

uvNikita commented:
See comment here: https://github.com/nix-community/home-manager/pull/3563#issuecomment-1373399177

GaetanLepage commented:
All right ! Thank you for the quick answer :)
I realize that with the current way the module is written, it is not possible to set the `default_[foating_]border` options to all possible settings described in [i3 documentation](https://i3wm.org/docs/userguide.html#default_border).
The value can be either `normal` or `pixel` but not `none` or `pixel 3` or `normal 42`.

uvNikita commented:
I think `pixel 3` should be possible to achieve with

```
window = {
  titlebar = false;
  border = 5;
}
```

GaetanLepage commented:
You are right, I had missed those :)

uvNikita commented:
As for the `none`, according to documentation:

```
# The same as default_border none
default_border pixel 0
```

So I think that's the reason of not including `none` into options.
